projects/plugins/crm/src/automation/commons/conditions/tags/class-object-tag.php

```diff
@@ -0,0 +1,261 @@
+			'has_tag'     => __( 'Has tag', 'zero-bs-crm' ),
+			'not_has_tag' => __( 'Does not have tag', 'zero-bs-crm' ),
+		);
+
+		$this->set_attribute_definitions(
+			array(
+				new Attribute_Definition( 'operator', __( 'Operator', 'zero-bs-crm' ), __( 'Determines how the field is compared to the specified value.', 'zero-bs-crm' ), Attribute_Definition::SELECT, $this->valid_operators ),
+				new Attribute_Definition( 'tag', __( 'Tag', 'zero-bs-crm' ), __( 'Object Tag to compare with.', 'zero-bs-crm' ), Attribute_Definition::TEXT ),
+			)
+		);
+	}
+
+	/**
+	 * Check for valid parameters.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param string $operator The operator.
+	 * @param mixed  $data The data to validate.
+	 * @param mixed  $previous_data The previous data to validate.
+	 * @return bool True if parameters are valid, false otherwise.
+	 */
+	private function check_for_valid_parameters( $operator, $data, $previous_data ) {
+
+		switch ( $operator ) {
+			case 'tag_added':
+			case 'tag_removed':
+				if ( $previous_data === null || ! $this->is_valid_tag_data( $data ) || ! $this->is_valid_tag_data( $previous_data ) ) {
+					$this->logger->log( 'Invalid tag data: parameters not valid' );
+
+					return false;
+				}
+				break;
+			case 'has_tag':
+			case 'not_has_tag':
+				if ( ! $this->is_valid_tag_data( $data ) ) {
+					$this->logger->log( 'Invalid tag data: parameters not valid' );
+					$this->condition_met = false;
+
+					return false;
+				}
+				break;
+			default:
+				$this->logger->log( 'Unknown operator: ' . $operator );
+				return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Checks if a given tag name exists in the tags array.
+	 *
+	 * @param array  $data     The array containing tag data.
+	 * @param string $tag_name The name of the tag to check for.
+	 *
+	 * @return bool True if the tag name exists, false otherwise.
+	 */
+	private function has_tag_by_name( $data, string $tag_name ) {
+		if ( ! is_array( $data ) ) {
+			$this->logger->log( 'The given tag data is not in the correct (array) format, so is not valid.' );
+			return false;
+		}
+		foreach ( $data as $item ) {
+			if ( is_array( $item ) && $item['name'] === $tag_name ) {
+				return true;
+			}
+		}
+		$this->logger->log( 'Tag does not exist in the given array of tags: ' );
+		return false;
+	}
+
+	/**
+	 * Executes and valid conditions. If the condition is met, the value stored in the
+	 * attribute $condition_met is set to true; otherwise, it is set to false.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param Data_Type $data Data passed from the trigger.
+	 * @return void
+	 *
+	 * @throws Automation_Exception If an invalid operator is encountered.
+	 */
+	public function execute( Data_Type $data ) {
+
+		$operator = $this->get_attributes()['operator'];
+		$tag      = $this->get_attributes()['tag'];
+
+		$this->check_for_valid_operator( $operator );
+
+		$previous_data = $data->get_previous_data();
+		$data          = $data->get_data();
```

Comment:
Better if we name both with a the `tag_` prefix rather than overwrite the `$data` variable.

Reply:
Or being more explicit: `$tag_list` and `$previous_tag_list`